Add the ability to auto-fix lint violations

LuKks opened the pull request:
Fix #13 
I never used Python so I strongly accept suggestions.

By the way, I did static the method `ensure_plugin_installed` so we don't move the entire code, anyways I think that is ok.


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#13: Auto-fix on save option](https://issuehunt.io/repos/40055697/issues/13)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

sindresorhus commented on linter.py:
```diff
@@ -73,3 +82,52 @@ def ensure_plugin_installed(self) -> bool:
+	if isinstance(data, str):
+		data = data.encode()
+
+	proc = subprocess.Popen(
```
Should use `subprocess.run()`.

LuKks replied:
Yes but Sublime Text is using Python v2, and `subprocess.run()` is available in Python v3.

LuKks replied:
For example, in `gofmt`:
https://github.com/noonat/sublime-gofmt/blob/master/gofmt.py#L93

sindresorhus replied:
We can drop Python 2 support. I'm not interested in supporting older Sublime versions anyway.

LuKks replied:
I have the latest version on Xubuntu 20.04:
![image](https://user-images.githubusercontent.com/12686176/82732625-ac92cd80-9ce4-11ea-9913-a4a1571e7d77.png)
The code doesn't work with `subprocess.run()`, so I checked the Python version and it said 2.7, I went to the documentation and `subprocess.run()` is available from 3.5.

Not sure what should I do but I have everything very default, so a lot of people has Sublime with Python 2, right?

When I researched about `cwd` I noticed that they use `subprocess.Popen()` as well.

sindresorhus replied:
Actually, Sublime 3 ships with Python 3.3: http://www.sublimetext.com/docs/3/porting_guide.html

So yes, it does indeed look like it's too old...

Can you add a TODO comment about using `.run()` when Sublime eventually upgrades to 3.5?

LuKks replied:
When Sublime get updated I can do the change, I will try to remember it or you can let me know.

sindresorhus replied:
I don't think it will happen anytime soon. I remember reading on the forums that there was a reason it was outdated.

sindresorhus commented on Default.sublime-commands:
```diff
@@ -0,0 +1,6 @@
+[
+  {
+    "caption": "SublimeLinter XO: Fix current file",
+    "command": "xofix"
```
```suggestion
    "command": "xo-fix"
```

LuKks replied:
I tried changing the `XofixCommand` to `XoFixCommand` but for some reason the command doesn't work anyways.
The command `xo_fix` worked, so we can use that one, it's ok?

By the way, I was just following the same convention as Go:
https://github.com/noonat/sublime-gofmt/blob/master/Default.sublime-commands#L4

sindresorhus replied:
> The command xo_fix worked, so we can use that one, it's ok?

👍

sindresorhus commented:
> I never used Python so I strongly accept suggestions.

Then I recommend you read up on Sublime Plugin and Python best practices. I have bounties on issues to reduce my workload, not increase it.

LuKks commented:
Just because I never used Python doesn't mean the code is wrong or something, I know other languages so I tried to make it very clean.
I read several Sublime Plugins before submitting the pull request.

sindresorhus commented on Main.sublime-menu:
```diff
@@ -0,0 +1,49 @@
+[
+  {
+  	"caption": "Tools",
```
Mixed indentation here. Just make everything tab-indented.

sindresorhus commented on Main.sublime-menu:
```diff
@@ -0,0 +1,49 @@
+								"caption": "Settings - Default",
+								"command": "open_file",
+								"args": {
+									"file": "${packages}/SublimeLinter-contrib-xo/SublimeLinterContribXO.sublime-settings"
```
Is there really no variable for the full path to the plugin or the sublime-settings file?

LuKks replied:
It seems not.
https://www.sublimetext.com/docs/3/build_systems.html#variables
http://gohom.win/ManualHom/IDE/Sublime/sublime-text-unofficial-documentation-3.0/#build-system-variables

sindresorhus commented on Main.sublime-menu:
```diff
@@ -0,0 +1,49 @@
+								"caption": "Settings - User",
+								"command": "open_file",
+								"args": {
+									"file": "${packages}/User/SublimeLinterContribXO.sublime-settings"
```
Same as the above question.

sindresorhus commented on SublimeLinterContribXO.sublime-settings:
```diff
@@ -0,0 +1,4 @@
+{
+	"cmd": ["xo", "--stdin", "--fix"],
```
Why does this need to be configurable? I would prefer it would be not.

sindresorhus commented on linter.py:
```diff
@@ -73,3 +82,52 @@ def ensure_plugin_installed(self) -> bool:
+	if view.file_name():
+		return os.path.dirname(view.file_name())
+	elif len(view.window().folders()):
+		return view.window().folders()[0]
```
Is there really no Sublime API to get the `cwd`?

LuKks replied:
Sublime let we code without creating a file or having multiple tabs with different locations, so the cwd becomes more tricky. Anyways, they should provide a simple way to get the `cwd` because this a common use case.
As I understood if we use `os.getcwd()` we get the `cwd` of Sublime installation.

With that in mind, previously I already followed gofmt:
https://github.com/noonat/sublime-gofmt/blob/master/gofmt.py#L48
And now I tried to research deeply and others are using very similar code.

LuKks commented:
It's ok?

sindresorhus commented:
When I run the "Fix current file", I get this error message in the Sublime console:

```
Traceback (most recent call last):
  File "/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 1088, in run_
    return self.run(edit)
  File "/Users/sindresorhus/Library/Application Support/Sublime Text 3/Packages/SublimeLinter-contrib-xo/linter.py", line 125, in run
    replacement = xo_fix(self.view, content)
  File "/Users/sindresorhus/Library/Application Support/Sublime Text 3/Packages/SublimeLinter-contrib-xo/linter.py", line 114, in xo_fix
    code = run_cmd(cmd, content, view)
  File "/Users/sindresorhus/Library/Application Support/Sublime Text 3/Packages/SublimeLinter-contrib-xo/linter.py", line 102, in run_cmd
    cwd=cwd,
  File "./python3.3/subprocess.py", line 819, in __init__
  File "./python3.3/subprocess.py", line 1448, in _execute_child
FileNotFoundError: [Errno 2] No such file or directory: 'xo'
```

I think `cmd = ('xo', '--stdin', '--reporter', 'compact', '--filename', '${file}')` does something special to put the local `xo` binary in the `PATH` and you don't get that for free when running the binary directly.

sindresorhus commented on readme.md:
```diff
@@ -23,6 +23,22 @@ In order for `xo` to be executed by SublimeLinter, you must ensure that its path
+
+## Auto-fix
+
+To run the auto-fixer, press the `ctrl+alt+f` shortcut (can be changed) or use the menu entry “Tools › SublimeLinter XO › Fix current file”.
```
Can you describe to the user how to change the shortcut?

LuKks commented:
I had have installed `xo` package globally so I didn't notice that error otherwise I wouldn't have pushed changes. Now it works :)

sindresorhus commented:
I noticed that many plugins that provide keyboard shortcuts have menu items for it. I think we should do the same:

<img width="392" alt="Screenshot 2020-06-11 at 13 26 12" src="https://user-images.githubusercontent.com/170270/84348443-837ca300-abe7-11ea-8d2c-9c7c1277d2c2.png">

sindresorhus commented:
I tried running the fix command now, and it just replaces the whole file with nothing.

LuKks commented:
It would be great if you can test it again, a friend mine tested it on Mac and it worked. I'm a bit busy right now so I will do the key bindings later.

sindresorhus commented:
I tried the latest commit now and it's still just emptying the file. No errors/warnings in the Sublime console. Happy to help you debug if you tell me what to do.

LuKks commented:
I added several prints/logs to basically know what is going, that should be enough.

My full procedure is:
1) Create a folder named `aa` on the desktop, then:
1.a) `npm init -y`
1.b) `npm i xo --save-dev`
1.c) Create a file named `index.js` with the content `var name = "lucas";`
2) On `Sublime Text` I installed `SublimeLinter` with `Package Control` (it's a requirement).
3) I moved to the next directory:
3.a) Linux: `cd ~/.config/sublime-text-3/Packages/`
3.b) OSX (should be): `cd ~/Library/Application\ Support/Sublime\ Text\ 3/Packages/`
4) `git clone -b fix-on-save https://github.com/LuKks/SublimeLinter-contrib-xo.git`
5) Close and open `Sublime Text` to clean logs.
6) Open the `index.js` previously created.
7) `View > Show console`, here in Linux I lose focus so click on the file or code.
8) `Tools > SublimeLinter XO > Fix current file`.

I hope that explanation is not overkill (?), anyways it's just install the package and test it.

In my case the logs are more or less:
```
XoFixCommand -> is_enabled?
XoFixCommand -> xo_start_dir /home/lucas/Desktop/aa
XoFixCommand -> xo_path /home/lucas/Desktop/aa/node_modules/.bin/xo
XoFixCommand -> return -> True
xo_fix -> content length: 20
xo_fix -> encoding: UTF-8
xo_fix -> xo_start_dir: /home/lucas/Desktop/aa
xo_fix -> xo_env.PWD: /home/lucas/Desktop/aa
xo_fix -> stdout: 22
```
And it obviously result in:
`const name = 'lucas';`

sindresorhus commented:
I just tested with https://github.com/chalk/chalk:

```
git: tracking working dir /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> is_enabled?
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> return -> True
XoFixCommand -> is_enabled?
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> return -> True
xo_fix -> content length: 6074
xo_fix -> encoding: UTF-8
xo_fix -> xo_start_dir: /Users/sindresorhus/dev/oss/chalk/source
xo_fix -> xo_env.PWD: /Users/sindresorhus/dev/oss/chalk/source
xo_fix -> stdout: 0
```

LuKks commented:
I think I got it, we can see that your real project root is `/Users/sindresorhus/dev/oss/chalk` but it's actually using the `xo_start_dir` `/Users/sindresorhus/dev/oss/chalk/source` as cwd which is wrong.

So, it would be awesome if you can test it again :)

LuKks commented:
I will push some changes where I removed the logs (hoping that anyway it's works) and I also added the `key bindings` into the package preferences.
In case you need it, you can still use the previous commit to see the logs.

sindresorhus commented:
I'm still seeing the same problem, unfortunately.

```
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> project_root -> /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> return -> True
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> project_root -> /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> return -> True
xo_fix -> content length: 6074
xo_fix -> encoding: UTF-8
xo_fix -> xo_start_dir: /Users/sindresorhus/dev/oss/chalk/source
xo_fix -> xo_env.PWD: /Users/sindresorhus/dev/oss/chalk
xo_fix -> xo_project_root: /Users/sindresorhus/dev/oss/chalk
xo_fix -> stdout: 0
```

Tested with commit https://github.com/xojs/SublimeLinter-contrib-xo/pull/21/commits/bf6eeb592d1407714b9cf8b85d94822fa700c210.

sindresorhus commented:
I'm also getting this, but probably not related:

> SublimeLinter: WARNING: xo: Defining 'cls.npm_name' has no effect. Please cleanup and remove this setting.

LuKks commented:
No way, that should have worked d:
Not sure what to do, I just reverted a commit to have back the logs and now I'm showing `stderr` just in case there is something there.
I'm starting to mistrust in `chalk`, can you test without it? Like a normal `sublime-package`?

```
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /home/lucas/Desktop/aaa/bbb
XoFixCommand -> xo_path /home/lucas/Desktop/aaa/node_modules/.bin/xo
XoFixCommand -> project_root -> /home/lucas/Desktop/aaa
XoFixCommand -> return -> True
xo_fix -> content length: 20
xo_fix -> encoding: UTF-8
xo_fix -> xo_start_dir: /home/lucas/Desktop/aaa/bbb
xo_fix -> xo_env.PWD: /home/lucas/Desktop/aaa
xo_fix -> xo_project_root: /home/lucas/Desktop/aaa
xo_fix -> stdout len: 22
xo_fix -> stderr len: 0
xo_fix -> stderr content: 
```

sindresorhus commented:
At least the problem is clear now:

```
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> project_root -> /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> return -> True
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> project_root -> /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> return -> True
xo_fix -> content length: 6074
xo_fix -> encoding: UTF-8
xo_fix -> xo_start_dir: /Users/sindresorhus/dev/oss/chalk/source
xo_fix -> xo_env.PWD: /Users/sindresorhus/dev/oss/chalk
xo_fix -> xo_project_root: /Users/sindresorhus/dev/oss/chalk
xo_fix -> stdout len: 0
xo_fix -> stderr len: 37
xo_fix -> stderr content: env: node: No such file or directory

xo_fix -> returncode: 127
```

I think you're missing out on some `node` binary resolution logic when spawning normally. Because normal linting works fine.

sindresorhus commented:
Also, if there's an error, you should not replace the contents, but instead show an error dialog.

sindresorhus commented:
It's also probably smart to keep all the logging for now even in when publishing the next release. In case there are other problems, it will be easier to debug.

LuKks commented:
Added a dialog when there is an error.

I'm also seeing about the enviroment (PATH) for MacOS, I think this is something common, look:
http://www.sublimelinter.com/en/latest/troubleshooting.html#debugging-path-problems
So, I think the normal solution is that you have to change your setting `linters -> eslint -> env`.

Anyways, it should work without changing any setting.
It would be awesome if you can test it again, this is my current output:
```
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /home/lucas/Desktop/aaa/bbb
XoFixCommand -> xo_path /home/lucas/Desktop/aaa/node_modules/.bin/xo
XoFixCommand -> shellenv -> shellenv.path -> /home/lucas/.deno/bin:/home/lucas/.npm-global/bin:/home/lucas/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/go/bin:/usr/local/node/bin
XoFixCommand -> environ  -> environ.path -> /home/lucas/.deno/bin:/home/lucas/.npm-global/bin:/home/lucas/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/go/bin:/usr/local/node/bin
XoFixCommand -> project_root -> /home/lucas/Desktop/aaa
XoFixCommand -> return -> True
xo_fix -> content length: 20
xo_fix -> encoding: UTF-8
xo_fix -> xo_start_dir: /home/lucas/Desktop/aaa/bbb
xo_fix -> xo_env.PWD: /home/lucas/Desktop/aaa
xo_fix -> xo_project_root: /home/lucas/Desktop/aaa
xo_fix -> stdout len: 23
xo_fix -> stderr len: 0
xo_fix -> stderr content: 
xo_fix -> returncode: 0
```

`which node`
`/usr/bin/node`

In Linux it was working but now we're using `shellenv` to get the correct enviroment for Mac, it's used in `sublime-gofmt`.
So, the main change is that now it's using the `shellenv.path` instead of the `environ.path` showed in the logs.

LuKks commented:
I tested that changes on Mac but `shellenv` is not working, I also tried `sublime-gofmt` and it's also not working due `shellenv`. So I think we can't use that library.

I reverted some code and I tested it on Mac and it's working just fine, maybe you have to check if your PATH for Node.js is correct like in the previous links related to `SublimeLinter`?

This is the actual output for Mac:
```
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/usuario/aaa
XoFixCommand -> xo_path /Users/usuario/aaa/node_modules/.bin/xo
XoFixCommand -> environ.path -> /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin
XoFixCommand -> project_root -> /Users/usuario/aaa
XoFixCommand -> return -> True
xo_fix -> content length: 19
xo_fix -> encoding: utf-8
xo_fix -> xo_env.PWD: /Users/usuario/aaa
xo_fix -> xo_project_root: /Users/usuario/aaa
xo_fix -> stdout len: 22
xo_fix -> stderr len: 130
xo_fix -> stderr content: (node:1612) ExperimentalWarning: Readable[Symbol.asyncIterator] is an experimental feature. This feature could change at any time

xo_fix -> returncode: 0
```

`which node`
`/usr/local/bin/node`

By the way, I improved the dialog to only show up when there is a real error. For example, in that output there is a warning, probably due the node version, but not an actual error so that's ok.

sindresorhus commented:
Nice! Works now.

---

For reference, here's the output:

```
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> environ.path -> /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/usr/local/opt/python@3.8/bin:/Users/sindresorhus/.gem/ruby/2.6.0/bin:/usr/local/lib/ruby/gems/2.7.0/bin:/Users/sindresorhus/.cargo/bin:/Users/sindresorhus/n/bin:/Users/sindresorhus/.dotfiles/bin:/usr/local/sbin:/opt/local/bin:/opt/local/sbin:/Users/sindresorhus/.dotfiles/completions:/usr/local/share/zsh-completions:/usr/local/share/zsh/site-functions:/usr/share/zsh/site-functions:/usr/share/zsh/5.7.1/functions:/usr/local/bin
XoFixCommand -> project_root -> /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> return -> True
XoFixCommand -> is_enabled?
XoFixCommand -> project_root -> None
XoFixCommand -> xo_start_dir /Users/sindresorhus/dev/oss/chalk/source
XoFixCommand -> xo_path /Users/sindresorhus/dev/oss/chalk/node_modules/.bin/xo
XoFixCommand -> environ.path -> /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/usr/local/opt/python@3.8/bin:/Users/sindresorhus/.gem/ruby/2.6.0/bin:/usr/local/lib/ruby/gems/2.7.0/bin:/Users/sindresorhus/.cargo/bin:/Users/sindresorhus/n/bin:/Users/sindresorhus/.dotfiles/bin:/usr/local/sbin:/opt/local/bin:/opt/local/sbin:/Users/sindresorhus/.dotfiles/completions:/usr/local/share/zsh-completions:/usr/local/share/zsh/site-functions:/usr/share/zsh/site-functions:/usr/share/zsh/5.7.1/functions:/usr/local/bin
XoFixCommand -> project_root -> /Users/sindresorhus/dev/oss/chalk
XoFixCommand -> return -> True
xo_fix -> content length: 6074
xo_fix -> encoding: UTF-8
xo_fix -> xo_env.PWD: /Users/sindresorhus/dev/oss/chalk
xo_fix -> xo_project_root: /Users/sindresorhus/dev/oss/chalk
xo_fix -> stdout len: 6076
xo_fix -> stderr len: 0
xo_fix -> stderr content: 
xo_fix -> returncode: 0
```
